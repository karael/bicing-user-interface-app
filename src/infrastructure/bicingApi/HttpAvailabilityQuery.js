import axios from 'axios';

class HttpAvailabilityQuery {
  static findAll() {
    return axios.get('http://localhost:8080/availabilities');
  }
}

export default HttpAvailabilityQuery;
