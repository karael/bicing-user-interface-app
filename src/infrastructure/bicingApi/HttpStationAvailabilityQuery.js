import axios from 'axios';
import { OK } from 'http-status';

import HttpStationQueryError from 'infrastructure/bicingApi/errors/HttpStationQueryError';

/** @todo make this class as a service (singleton) as pass API_URL threw construction */
class HttpStationAvailabilityQuery {
  static find(stationId, periodStart, periodEnd, interval) {
    return new Promise((resolve, reject) => {
      axios
        .get(`${process.env.REACT_APP_BICING_API_URL}/stations/${stationId}/availabilities`)
        .then(response => {
          if (OK !== response.status) {
            reject({
              error: HttpStationQueryError.withUnexpectedResponseStatus(response.status)
            });
          }
          console.log(response.data);
          // if () @todo add test to check if response.data
          resolve(response.data);
        }).catch(error => {
          reject({
            error: HttpStationQueryError.withRequestError(error)
          });
        })
    });
  }
}

export default HttpStationAvailabilityQuery;
