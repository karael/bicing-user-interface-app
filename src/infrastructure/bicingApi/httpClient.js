import axios from 'axios';

const httpClient = axios.create({
  baseURL: `${process.env.REACT_APP_BICING_API_URL}`,
});

export default httpClient;
