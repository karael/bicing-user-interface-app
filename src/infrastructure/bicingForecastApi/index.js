import HttpStationAvailabilitiesForecastQuery from 'infrastructure/bicingForecastApi/HttpStationAvailabilitiesForecastQuery';

export default {
  HttpStationAvailabilitiesForecastQuery,
};
