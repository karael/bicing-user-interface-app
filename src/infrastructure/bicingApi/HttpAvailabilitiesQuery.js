import { OK } from 'http-status';
import HttpAvailabilitiesQueryError
  from 'infrastructure/bicingApi/errors/HttpAvailabilitiesQueryError';
import httpClient from 'infrastructure/bicingApi/httpClient';
import httpHydraCollectionResponse
  from 'infrastructure/bicingApi/types/httpHydraCollectionResponse';

class HttpAvailabilitiesQuery {
  static async find() {
    const apiResponse = await httpClient.get('/last-availabilities-by-station')
      .then(response => response)
      .catch((error) => {
        throw HttpAvailabilitiesQueryError.withRequestError(error);
      });

    if (OK !== apiResponse.status) {
      throw HttpAvailabilitiesQueryError.withUnexpectedResponseStatus(apiResponse.status);
    }

    return HttpAvailabilitiesQuery.validate(apiResponse.data);
  }

  static validate(data) {
    return httpHydraCollectionResponse.validate(data)
      .then(result => (result['hydra:member']))
      .catch((validationError) => {
        throw HttpAvailabilitiesQueryError.withResponseFormatValidationErrors(
          validationError.details.map(d => d.message),
        );
      });
  }
}

export default HttpAvailabilitiesQuery;
