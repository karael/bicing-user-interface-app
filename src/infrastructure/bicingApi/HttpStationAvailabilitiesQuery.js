import ByIntervalInPeriodFilter from 'application/state/filter/ByIntervalInPeriodFilter';
import { OK } from 'http-status';

import HttpStationQueryError from 'infrastructure/bicingApi/errors/HttpStationQueryError';
import httpClient from 'infrastructure/bicingApi/httpClient';

class HttpStationAvailabilitiesQuery {
  static async find(stationId, byFilter) {
    const apiResponse = await httpClient
      .get(HttpStationAvailabilitiesQuery.uri(stationId, byFilter))
      .then(response => response)
      .catch((error) => {
        throw HttpStationQueryError.withRequestError(error);
      });

    if (OK !== apiResponse.status) {
      throw HttpStationQueryError.withUnexpectedResponseStatus(apiResponse.status);
    }

    return apiResponse.data;
  }

  static uri(stationId, byFilter) {
    const uri = `/stations/${stationId}/availabilities`;

    if (byFilter instanceof ByIntervalInPeriodFilter) {
      return `${uri}?periodStart=${byFilter.periodStart}&periodEnd=${byFilter.periodEnd}&interval=${byFilter.interval}`;
    }

    return uri;
  }
}

export default HttpStationAvailabilitiesQuery;
