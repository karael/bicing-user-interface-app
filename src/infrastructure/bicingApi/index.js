import HttpAvailabilitiesQuery from 'infrastructure/bicingApi/HttpAvailabilitiesQuery';
import HttpStationAvailabilitiesQuery
  from 'infrastructure/bicingApi/HttpStationAvailabilitiesQuery';
import HttpStationQuery from 'infrastructure/bicingApi/HttpStationQuery';
import HttpStationsQuery from 'infrastructure/bicingApi/HttpStationsQuery';

export {
  HttpAvailabilitiesQuery,
  HttpStationAvailabilitiesQuery,
  HttpStationQuery,
  HttpStationsQuery,
};
