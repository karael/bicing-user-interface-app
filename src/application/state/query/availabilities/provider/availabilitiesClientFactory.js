import ByItineraryAtFilter from 'application/state/filter/ByItineraryAtFilter';
import { HttpAvailabilitiesQuery } from 'infrastructure/bicingApi';
import { HttpAvailabilitiesForecastQuery } from 'infrastructure/bicingForecastApi';

const availabilitiesClientFactory = async (
  itineraryAt,
  periodStart,
  periodEnd,
  interval,
  stationIds) => {
  const filter = await ByItineraryAtFilter.fromRawValues(
    itineraryAt,
    periodStart,
    periodEnd,
    interval,
  );

  return filter.isForecasting() === true
    ? new HttpAvailabilitiesForecastQuery(stationIds, filter)
    : new HttpAvailabilitiesQuery();
};

export default availabilitiesClientFactory;
