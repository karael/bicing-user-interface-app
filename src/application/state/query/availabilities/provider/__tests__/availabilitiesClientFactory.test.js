import ByIntervalInPeriodFilter from 'application/state/filter/ByIntervalInPeriodFilter';
import availabilitiesClientFactory
  from 'application/state/query/availabilities/provider/availabilitiesClientFactory';
import { HttpAvailabilitiesQuery } from 'infrastructure/bicingApi';
import { HttpAvailabilitiesForecastQuery } from 'infrastructure/bicingForecastApi';
import { v4 as uuid } from 'uuid';

describe('application/state/query/availabilities/provider/availabilitiesClientFactory', () => {
  test('should create a default client if no filter parameters', () => {
    const client = availabilitiesClientFactory(null, null, null, null);

    expect(client).toBeInstanceOf(HttpAvailabilitiesQuery);
    expect(client).toEqual(new HttpAvailabilitiesQuery());
  });
  test('should create a specific client if filter parameters', () => {
    const periodStart = '2017-12-12 12:12:12';
    const periodEnd = '2017-12-12 12:52:12';
    const interval = '5T';
    const stationIds = [uuid(), uuid(), uuid()];
    const filter = ByIntervalInPeriodFilter.fromRawValues(periodStart, periodEnd, interval);

    const client = availabilitiesClientFactory(periodStart, periodEnd, interval, stationIds);

    expect(client).toBeInstanceOf(HttpAvailabilitiesForecastQuery);
    expect(client.stationIds).toEqual(stationIds);
    expect(client.byFilter).toEqual(filter);
  });
  test('should not create a specific client if filter parameters are bad and should throw an error', () => {
    expect(() => availabilitiesClientFactory(
      'bad_value',
      '2017-12-12 12:52:12',
      '5T',
      [uuid(), uuid(), uuid()],
    )).toThrowErrorMatchingSnapshot();
  });
});
