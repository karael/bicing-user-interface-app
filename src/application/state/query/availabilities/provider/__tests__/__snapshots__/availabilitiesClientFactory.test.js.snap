// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`application/state/query/availabilities/provider/availabilitiesClientFactory should not create a specific client if filter parameters are bad and should throw an error 1`] = `"Invalid date format"`;
